src/hotspot/share/opto/mempointer.hpp

```diff
@@ -461,98 +536,276 @@ class MemPointerSummand : public StackObj {
+// "inner" nodes of the pointer expression. This callback is for example
+// used in SuperWord::unrolling_analysis to collect all inner nodes of a
+// pointer expression.
+class MemPointerParserCallback : public StackObj {
+private:
+  static MemPointerParserCallback _empty;
+
+public:
+  virtual void callback(Node* n) { /* do nothing by default */ }
+
+  // Singleton for default arguments.
+  static MemPointerParserCallback& empty() { return _empty; }
+};
+
 // Decomposed form of the pointer sub-expression of "pointer".
 //
 //   pointer = SUM(summands) + con
 //
-class MemPointerDecomposedForm : public StackObj {
-private:
+// Node: if the base is known, then it is in the 0th summand. A base can be:
+//       - on-heap  / object: base().object()
+//       - off-heap / native: base().native()
+class MemPointer : public StackObj {
+public:
   // We limit the number of summands to 10. This is just a best guess, and not at this
   // point supported by evidence. But I think it is reasonable: usually, a pointer
   // contains a base pointer (e.g. array pointer or null for native memory) and a few
   // variables. It should be rare that we have more than 9 variables.
   static const int SUMMANDS_SIZE = 10;
 
-  Node* _pointer; // pointer node associated with this (sub)pointer
+  class Base : public StackObj {
+  private:
+    enum Kind { Unknown, Object, Native };
+    Kind _kind;
+    Node* _base;
 
-  MemPointerSummand _summands[SUMMANDS_SIZE];
-  NoOverflowInt _con;
+    Base(Kind kind, Node* base) : _kind(kind), _base(base) {
+      assert((kind == Unknown) == (base == nullptr), "known base");
+    }
 
-public:
-  // Empty
-  MemPointerDecomposedForm() : _pointer(nullptr), _con(NoOverflowInt::make_NaN()) {}
+  public:
+    Base() : Base(Unknown, nullptr) {}
+    static Base make(Node* pointer, const GrowableArray<MemPointerSummand>& summands);
+
+    bool is_known()          const { return _kind != Unknown; }
+    bool is_object()         const { return _kind == Object; }
+    bool is_native()         const { return _kind == Native; }
+    Node* object()           const { assert(is_object(), ""); return _base; }
+    Node* native()           const { assert(is_native(), ""); return _base; }
+    Node* object_or_native() const { assert(is_known(),  ""); return _base; }
+    Node* object_or_native_or_null() const { return _base; }
+
+#ifndef PRODUCT
+    void print_on(outputStream* st) const {
+      switch (_kind) {
+      case Object:
+          st->print("object  ");
+          st->print("%d %s", _base->_idx, _base->Name());
+          break;
+      case Native:
+          st->print("native  ");
+          st->print("%d %s", _base->_idx, _base->Name());
+          break;
+      default:
+          st->print("unknown ");
+      };
+    }
+#endif
+
+  private:
+    static Node* find_base(Node* object_base, const GrowableArray<MemPointerSummand>& summands);
+  };
 
 private:
+  NOT_PRODUCT( const TraceMemPointer& _trace; )
```

Comment:
Why you prefer `_trace` to be first and not last?

Reply:
Ah, I think in a previous version I needed the _trace while initializing other parts... but I don't any more.

Nice catch!